File: src/main/voice/tts.ts
import { exec } from "child_process";

export function speak(text: string): Promise<void> {
  return new Promise((resolve) => {
    // Sanitize text to prevent command injection
    const sanitized = text.replace(/[\\"`$]/g, "");

    if (process.platform === "darwin") {
      exec(`say "${sanitized}"`, () => resolve());
    } else if (process.platform === "win32") {
      const escaped = sanitized.replace(/'/g, "''");
      exec(
        `powershell -Command "Add-Type -AssemblyName System.Speech; (New-Object System.Speech.Synthesis.SpeechSynthesizer).Speak('${escaped}')"`,
        () => resolve()
      );
    } else {
      // Linux fallback: espeak
      exec(`espeak "${sanitized}"`, () => resolve());
    }
  });
}

File: src-tauri/src/tts.rs
use std::process::Command;

pub fn speak(text: &str) {
    let sanitized: String = text.chars().filter(|c| *c != '"' && *c != '\\' && *c != '$').collect();

    #[cfg(target_os = "macos")]
    {
        let _ = Command::new("say").arg(&sanitized).spawn();
    }

    #[cfg(target_os = "windows")]
    {
        let escaped = sanitized.replace('\'', "''");
        let _ = Command::new("powershell")
            .args(["-Command", &format!(
                "Add-Type -AssemblyName System.Speech; (New-Object System.Speech.Synthesis.SpeechSynthesizer).Speak('{}')",
                escaped
            )])
            .spawn();
    }

    #[cfg(target_os = "linux")]
    {
        let _ = Command::new("espeak").arg(&sanitized).spawn();
    }
}
